### PR TITLE
Dev transform

### PR DESCRIPTION
**Pull request by punith300i**
- Implemented Tranformation API which supports the following transformation types, by leveraging restrictedpython.

  - map
  - filter
  - split
  - concatenate

- Added unit test cases.

Screenshot 2023-08-22 at 12 43 03 PM

  Screenshot 2023-08-22 at 3 23 12 PM
  Screenshot 2023-08-22 at 3 23 53 PM
  Screenshot 2023-08-22 at 3 23 29 PM

- Added context variable for the user to have complete access of the row that belongs to a cell on which the transformation is applied.

### CODE REVIEW

File: sand/controllers/transform.py [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+from werkzeug.exceptions import BadRequest
+
+from sand.models.table import Link, Table, TableRow
+from sand.models.transform import Context
+
+transform_bp = Blueprint("transform", "transform")
+USER_DEFINE_FUNCTION_NAME = ""
+
+
+def transform_map(transform_func, data, tolerance):
```
Comment (binh-vu) [inline sand/controllers/transform.py:19]: add type hint

File: sand/controllers/transform.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    exec_tolerance = tolerance
+    for path, value, context in data:
+        value = value[0]
+        tdata = dict()
+        tdata["path"] = path
+        tdata["value"] = value
+        try:
+            result = transform_func(value, context)
+            tdata["ok"] = result
+            transformed_data.append(tdata)
```
Comment (binh-vu) [inline sand/controllers/transform.py:30]: duplicated with line 35. resolve this by changing return statement inside the loop to break

File: sand/controllers/transform.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+from sand.models.table import Link, Table, TableRow
+from sand.models.transform import Context
+
+transform_bp = Blueprint("transform", "transform")
+USER_DEFINE_FUNCTION_NAME = ""
+
+
+def transform_map(transform_func, data, tolerance):
+    transformed_data = []
+    exec_tolerance = tolerance
```
Comment (binh-vu) [inline sand/controllers/transform.py:21]: why reassign the variable?

File: sand/controllers/transform.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        value = value[0]
+        tdata = dict()
+        tdata["path"] = path
+        tdata["value"] = value
+        try:
+            result = transform_func(value, context)
+            tdata["ok"] = result
+            transformed_data.append(tdata)
+        except Exception as err:
+            (exc, value, tb) = sys.exc_info()
```
Comment (binh-vu) [inline sand/controllers/transform.py:32]: line 32-34 are duplicated, consider moving to a separate function

File: sand/controllers/transform.py [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+    transformed_data = []
+    exec_tolerance = tolerance
+    for path, value, context in data:
+        tdata = dict()
+        value = value[0]
+        tdata["path"] = path
+        tdata["value"] = value
+        try:
+            result = transform_func(value, context)
+            if type(result) != bool:
```
Comment (binh-vu) [inline sand/controllers/transform.py:52]: don't do type check like this, use instance of

File: sand/controllers/transform.py [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    transformed_data = []
+    exec_tolerance = tolerance
+    for path, value, context in data:
+        tdata = dict()
+        value = value[0]
+        tdata["path"] = path
+        tdata["value"] = value
+        try:
+            result = transform_func(value, context)
+            if type(result) != list:
```
Comment (binh-vu) [inline sand/controllers/transform.py:77]: do we expect the return type to be a list or iterable?

File: sand/controllers/transform.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+                raise BadRequest("filter transform function must return list")
```
Comment (binh-vu) [inline sand/controllers/transform.py:78]: bad msg

File: sand/controllers/transform.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+)
+def transform(table_id: int):
+    table = Table.get_by_id(table_id)
+    table_rows: List[TableRow] = list(
+        TableRow.select().where(TableRow.table == table).order_by(TableRow.index)
+    )
+    transform_type = request.json["type"]
+    mode = request.json["mode"]
+    datapath = request.json["datapath"]
+    if type(datapath) != list and type(datapath) == str:
```
Comment (binh-vu) [inline sand/controllers/transform.py:125]: this line is problematic

File: sand/controllers/transform.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+@transform_bp.route(
+    f"/{transform_bp.name}/<table_id>/transformations", methods=["POST"]
+)
+def transform(table_id: int):
+    table = Table.get_by_id(table_id)
+    table_rows: List[TableRow] = list(
+        TableRow.select().where(TableRow.table == table).order_by(TableRow.index)
+    )
+    transform_type = request.json["type"]
```
Comment (binh-vu) [inline sand/controllers/transform.py:122]: I think we discussed about validating the inputs in one of the meeting, so we need to do it

File: sand/controllers/transform.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+    tolerance = request.json["tolerance"]
+    outputpath = None
+    if "outputpath" in request.json:
+        outputpath = request.json["outputpath"]
+    rows = request.json["rows"]
+
+    loc = {}
+
+    # overriding inbuilt _getitem_ guard function
+    def custom_getitem_guard(iterable, index):
```
Comment (binh-vu) [inline sand/controllers/transform.py:138]: you seems don't understand what iterable means...

File: sand/controllers/transform.py [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+        return iterable[index]
+
+    # update safe_globals to enable _getitem_
+    safe_globals.update({'_getitem_': custom_getitem_guard})
```
Comment (binh-vu) [inline sand/controllers/transform.py:142]: the function compilation gets complicated. move to a separate function

File: sand/controllers/transform.py [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+    if compiled_result.errors:
+        raise BadRequest("\n".join(compiled_result.errors))
+
+    exec(compiled_result.code, safe_globals, loc)
+
+    transform_func = loc["<function>"]
+
+    col_index_list = [table.columns.index(column) for column in datapath]
+
+    data = [[table_row.index, [table_row.row[col_index] for col_index in col_index_list],
```
Comment (binh-vu) [inline sand/controllers/transform.py:154]: we aren't doing iterator anymore?

File: sand/models/transform.py [hunk 13]
```diff
@@ -0,0 +1,8 @@
+from dataclasses import dataclass
+
+
+@dataclass
+class Context:
```
Comment (binh-vu) [inline sand/models/transform.py:5]: some descriptions would be nice

File: tests/test_transform_api.py [hunk 14]
```diff
@@ -0,0 +1,519 @@
+def test_api_transform_map_single_line(client, example_db):
+    resp = client.post(
+        "/api/transform/1/transformations",
+        json={
+            "type": "map",
+            "mode": "restrictedpython",
+            "datapath": ["Tên"],
+            "code": "return value.upper()",
+            "tolerance": 3,
+            "rows": 5,
```
Comment (binh-vu) [inline tests/test_transform_api.py:10]: row = 5 but it returns so many data

File: tests/test_transform_api.py [hunk 15]
```diff
@@ -0,0 +1,519 @@
...
+                        {'ok': 'Ngok Phan', 'path': 20, 'value': ['Ngok Phan', '']},
+                        {'ok': 'Ngok Lum Heo', 'path': 21, 'value': ['Ngok Lum Heo', '']},
+                        {'ok': 'Ngok Krinh', 'path': 22, 'value': ['Ngok Krinh', '']}]
+    response = resp.json
+    assert resp.status_code == 200
+    assert len(response) == len(transformed_data)
+    assert response == transformed_data
+
+
+def test_api_transform_compilation_error(client, example_db):
```
Comment (binh-vu) [inline tests/test_transform_api.py:451]: add test case for restrictedpython doesn't allow getitem and your solution can handle it.

**Review by binh-vu — CHANGES_REQUESTED** [review]

File: sand/controllers/transform.py (hunk 5 above)
Comment (punith300i) [reply sand/controllers/transform.py:52]: Updated it, thanks

File: sand/controllers/transform.py (hunk 7 above)
Comment (punith300i) [reply sand/controllers/transform.py:78]: updated the chanages

File: sand/controllers/transform.py (hunk 8 above)
Comment (punith300i) [reply sand/controllers/transform.py:125]: fixed by using isinstance()

File: sand/controllers/transform.py (hunk 11 above)
Comment (punith300i) [reply sand/controllers/transform.py:142]: implemented compile_function()

File: sand/controllers/transform.py (hunk 9 above)
Comment (punith300i) [reply sand/controllers/transform.py]: Completed!

File: tests/test_transform_api.py (hunk 15 above)
Comment (punith300i) [reply tests/test_transform_api.py:467]: Added the test case!

File: tests/test_transform_api.py (hunk 14 above)
Comment (punith300i) [reply tests/test_transform_api.py]: updated the code to handle rows

File: sand/models/transform.py (hunk 13 above)
Comment (punith300i) [reply sand/models/transform.py:7]: added google style code

File: pyproject.toml [hunk 16]
```diff
@@ -15,7 +15,7 @@ include = ["sand/www/**/*"]
 sand = 'sand.__main__:cli'
 
 [tool.poetry.dependencies]
-python = ">=3.8,<4.0"
+python = ">=3.8,<3.12"
```
Comment (binh-vu) [inline pyproject.toml:18]: why restrict to 3.12?

File: sand/controllers/transform.py [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+from flask.blueprints import Blueprint
+from RestrictedPython import compile_restricted_function, safe_globals
+from werkzeug.exceptions import BadRequest
+
+from sand.models.table import Link, Table, TableRow
+from sand.models.transform import Context, Tdata, TransformRequestPayload
+from gena.deserializer import get_dataclass_deserializer
+
+transform_bp = Blueprint("transform", "transform")
+deserializer = get_dataclass_deserializer(TransformRequestPayload, {})
```
Comment (binh-vu) [inline sand/controllers/transform.py:17]: make this variable name more specific. transform_request_payload_deserializer is not great but better than deserializer. you can pick a better name

File: sand/controllers/transform.py [hunk 18]
```diff
@@ -0,0 +1,245 @@
...
+
+        Returns:
+            Error String without the sensitive information.
+    """
+    (exc, value, tb) = sys.exc_info()
+    tb = tb.tb_next
+    return "".join(traceback.format_exception(exc, value, tb))
+
+
+def transform_map(transform_func: Callable[[Any, Context], Any], data: Iterable[Tuple[int, List, Context]],
```
Comment (binh-vu) [inline sand/controllers/transform.py:33]: `Iterable[Tuple[int, List, Context]]`: in order to make this more useful, create a variable call ItemIndex = int, Item = Any, and use it in tuple[ItemIndex, Item, Context].

File: sand/controllers/transform.py [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+    table = Table.get_by_id(table_id)
+    table_rows: List[TableRow] = list(
+        TableRow.select().where(TableRow.table == table).order_by(TableRow.index)
+    )
+
+    if isinstance(request.json["datapath"], str):
+        request.json["datapath"] = [request.json["datapath"]]
+
+    if "rows" not in request.json:
+        request.json["rows"] = len(table_rows)
```
Comment (binh-vu) [inline sand/controllers/transform.py:209]: this is not necessary as a[:None] is = a[:].

File: sand/controllers/transform.py [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+        if request_data.outputpath:
+            if len(request_data.outputpath) != 1:
+                raise BadRequest(
+                    "For transform type map the outputpath should be a single column"
+                )
+        transformed_data = transform_map(transform_func, data, request_data.tolerance)
+
+    elif request_data.type == "filter":
+        if request_data.outputpath:
+            if len(request_data.outputpath) != 1:
```
Comment (binh-vu) [inline sand/controllers/transform.py:229]: don't do nested if like this.

File: sand/models/transform.py [hunk 21]
```diff
@@ -0,0 +1,29 @@
+from dataclasses import dataclass, InitVar
+from typing import Optional, Union, List, Iterable, Literal
+from typing_extensions import NotRequired, TypedDict
+
+
+@dataclass
+class Context:
+    """ Context dataclass to access the row of the cell that is being transformed."""
+    index: int
+    row: list
```
Comment (binh-vu) [inline sand/models/transform.py:10]: checkout the type of row in sand.models.table.TableRow

**Review by binh-vu — CHANGES_REQUESTED** [review]

File: sand/controllers/transform.py [hunk 22]
```diff
@@ -30,7 +38,11 @@ def filter_traceback_errors() -> str:
     return "".join(traceback.format_exception(exc, value, tb))
 
 
-def transform_map(transform_func: Callable[[Any, Context], Any], data: Iterable[Tuple[int, List, Context]],
+Item = Any
+ItemIndex = int
+
+
+def transform_map(transform_func: Callable[[Any, Context], Any], data: Tuple[ItemIndex, Item, Context],
```
Comment (binh-vu) [inline sand/controllers/transform.py:45]: data: Tuple[ItemIndex, Item, Context] where is Iterable?

File: sand/controllers/transform.py [hunk 23]
```diff
@@ -217,16 +226,14 @@ def transform(table_id: int):
     transformed_data = None
 
     if request_data.type == "map":
-        if request_data.outputpath:
-            if len(request_data.outputpath) != 1:
+        if request_data.outputpath and len(request_data.outputpath) != 1:
                 raise BadRequest(
                     "For transform type map the outputpath should be a single column"
                 )
         transformed_data = transform_map(transform_func, data, request_data.tolerance)
```
Comment (binh-vu) [inline sand/controllers/transform.py:233]: did you run unit test before pushing the code ???

**Review by binh-vu — CHANGES_REQUESTED** [review]

File: sand/controllers/transform.py (hunk 23 above)
Comment (punith300i) [reply sand/controllers/transform.py]: I added a new test case for it

File: sand/controllers/transform.py (hunk 22 above)
Comment (punith300i) [reply sand/controllers/transform.py]: updated with iterable

**Review by binh-vu — COMMENTED** [review]
forgot to delete search file

**Pull request merged**